### PR TITLE
Issue #1014

### LINKED ISSUE #1014: Validator Caching Race Condition
**Describe the bug**
Reading and writing to the validator map concurrently is possible. Use a RW lock to prevent this:

```
error: concurrent map read and map write

goroutine 1538904 [running]:
runtime.throw(0x107fdc5, 0x21)
	/usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc00daa2db0 sp=0xc00daa2d80 pc=0x433e42
runtime.mapaccess2(0xef8f40, 0xc000014e10, 0xc00daa2e20, 0xc00886ae70, 0x28)
	/usr/local/go/src/runtime/map.go:470 +0x278 fp=0xc00daa2df8 sp=0xc00daa2db0 pc=0x412968
github.com/pokt-network/pocket-core/x/nodes/keeper.Keeper.getValidatorFromCache(0x13327e0, 0xc0000136b0, 0xc0003819d0, 0x1357b20, 0xc000380380, 0x0, 0x0, 0xc00012c2a0, 0x13327e0, 0xc00014e3a0, ...)


```

### PR DESCRIPTION
**Pull request by andrewnguyen22**
closes #1014

### CODE REVIEW

**Review by oten91 — DISMISSED** [review]

File: app/cmd/cli/root.go [hunk 1]
```diff
@@ -50,16 +50,16 @@ func init() {
 	rootCmd.PersistentFlags().StringVar(&remoteCLIURL, "remoteCLIURL", "", "takes a remote endpoint in the form of <protocol>://<host> (uses RPC Port)")
 	rootCmd.PersistentFlags().StringVar(&persistentPeers, "persistent_peers", "", "a comma separated list of PeerURLs: '<ID>@<IP>:<PORT>,<ID2>@<IP2>:<PORT>...<IDn>@<IPn>:<PORT>'")
 	rootCmd.PersistentFlags().StringVar(&seeds, "seeds", "", "a comma separated list of PeerURLs: '<ID>@<IP>:<PORT>,<ID2>@<IP2>:<PORT>...<IDn>@<IPn>:<PORT>'")
-	rootCmd.Flags().BoolVar(&simulateRelay, "simulateRelay", false, "would you like to be able to test your relays")
-	rootCmd.Flags().BoolVar(&keybase, "keybase", true, "run wiith keybase, if disabled allows you to stake for the current validator only. providing a keybase is still neccesary for staking for apps & sending transactions")
+	startCmd.Flags().BoolVar(&simulateRelay, "simulateRelay", false, "would you like to be able to test your relays")
+	startCmd.Flags().BoolVar(&keybase, "keybase", true, "run wiith keybase, if disabled allows you to stake for the current validator only. providing a keybase is still neccesary for staking for apps & sending transactions")
```
Comment (luyzdeleon) [inline app/cmd/cli/root.go]: Fix `wiith` typo.

File: x/nodes/keeper/valStateChanges.go [hunk 2]
```diff
@@ -343,18 +343,26 @@ func (k Keeper) JailValidator(ctx sdk.Ctx, addr sdk.Address) {
...
 	validator.Jailed = true
 	k.SetValidator(ctx, validator)
 	logger := k.Logger(ctx)
 	logger.Info(fmt.Sprintf("validator %s jailed", addr))
+	ctx.EventManager().EmitEvent(
+		sdk.NewEvent(
+			types.EventTypeJail,
+			sdk.NewAttribute(types.AttributeKeyAddress, addr.String()),
+			sdk.NewAttribute(types.AttributeKeyReason, types.AttributeValueMissingSignature),
+			sdk.NewAttribute(types.AttributeKeyJailed, addr.String()),
```
Comment (luyzdeleon) [inline x/nodes/keeper/valStateChanges.go]: Shouldn't this be `validator.Jailed` instead of `addr.String()`

File: x/nodes/types/signing_info.go [hunk 3]
```diff
@@ -9,25 +9,28 @@ import (
...
+	Address             sdk.Address `json:"address" yaml:"address"`                             // validators address
+	StartHeight         int64       `json:"start_height" yaml:"start_height"`                   // height at which validator was first a candidate
+	Index               int64       `json:"index_offset" yaml:"index_offset"`                   // index offset into signed block bit array
 	JailedUntil         time.Time   `json:"jailed_until" yaml:"jailed_until"`                   // timestamp validator cannot be unjailed until
-	Tombstoned          bool        `json:"tombstoned" yaml:"tombstoned"`                       // whether or not a validator has been tombstoned (killed out of validator set)
 	MissedBlocksCounter int64       `json:"missed_blocks_counter" yaml:"missed_blocks_counter"` // missed blocks counter (to avoid scanning the array every time)
 	JailedBlocksCounter int64       `json:"jailed_blocks_counter" yaml:"jailed_blocks_counter"` // jailed blocks counter (to avoid scanning the array every time)
 }
 
+func (i *ValidatorSigningInfo) Reset() {
```
Comment (luyzdeleon) [inline x/nodes/types/signing_info.go:20]: What about resetting `StartHeight`?

File: x/nodes/types/signing_info.go [hunk 4]
```diff
@@ -9,25 +9,28 @@ import (
...
 type ValidatorSigningInfo struct {
-	Address             sdk.Address `json:"address" yaml:"address"`                             // validator consensus address
-	StartHeight         int64       `json:"start_height" yaml:"start_height"`                   // height at which validator was first a candidate OR was unjailed
-	IndexOffset         int64       `json:"index_offset" yaml:"index_offset"`                   // index offset into signed block bit array
+	Address             sdk.Address `json:"address" yaml:"address"`                             // validators address
+	StartHeight         int64       `json:"start_height" yaml:"start_height"`                   // height at which validator was first a candidate
+	Index               int64       `json:"index_offset" yaml:"index_offset"`                   // index offset into signed block bit array
 	JailedUntil         time.Time   `json:"jailed_until" yaml:"jailed_until"`                   // timestamp validator cannot be unjailed until
-	Tombstoned          bool        `json:"tombstoned" yaml:"tombstoned"`                       // whether or not a validator has been tombstoned (killed out of validator set)
```
Comment (luyzdeleon) [inline x/nodes/types/signing_info.go:16]: Why is this property being removed?

**Review by luyzdeleon — CHANGES_REQUESTED** [review]

File: app/cmd/cli/root.go (hunk 1 above)
Comment (andrewnguyen22) [reply app/cmd/cli/root.go]: Will fix

File: x/nodes/types/signing_info.go (hunk 3 above)
Comment (andrewnguyen22) [reply x/nodes/types/signing_info.go:20]: No, start height should remain the same for future tracking purposes

File: x/nodes/types/signing_info.go (hunk 4 above)
Comment (andrewnguyen22) [reply x/nodes/types/signing_info.go:16]: Never used

File: x/nodes/keeper/valStateChanges.go (hunk 2 above)
Comment (andrewnguyen22) [reply x/nodes/keeper/valStateChanges.go]: Maybe, this was copied over from legacy cosmos code

**Pull request merged**